docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+a limited capacity - whenever a new change comes from etcd and the cache is full,
+the oldest change will be removed to make place for the new change.
+
+When a client sends a watch request to apiserver, instead of redirecting it to
+etcd, it will just be registered in that cache. This will result in creating a
+go-routine that will be responsible for iterating over the cache, applying
+the matcher and sending filtered changes directly to the client.
+
+The following section describes the proposal in more details, analizes some
+corner cases and divides the whole design in more fine-grained steps.
+
+
+## Proposal details
+
+We would like the cache to be __per-resource-type__ and __optional__. Thanks to
+it we will be able to:
+  - have different cache sizes for different resources (e.g. bigger cache
+    [= longer history] for pods, which can significantly affect performance)
+  - avoid any overhead for objects that are watched very rarely (e.g. events
+    are almost not watched at all, but there are a lot of them)
+  - filter the cache for each watcher more effectively
+
+If we decide to support watches spanning different resources in the future and
+we have an efficient indexing mechanisms, it should be relatively simple to unify
+the cache to be common for all the resources.
+
+Since we want the watch in apiserver to be optional for different resource
+types, this needs to be self-contained and hidden behind a well defined API.
+It seems that this API can be very simple - it should just expose a method
+for initializing the cache and another one for registering (and starting) a
+new watcher with a given filtering function (matcher).
+
+The rest of this section describes the concrete steps that need to be done
+to implement the proposal.
+
+1. Since we are initially targeting to have per-resource-type watch in
+apiserver, this should probably be incorporated into generic etcd registry:
+"pkg/registry/generic/etcd". Thus we should migrate all the resources to
+use it (it seems it's not the case now e.g. for services).
+
+2. We should define an API for the watch in apiserver (this will be purely
+internal api just within apiserver, so it should be possible to change it
+relatively easy in the future). Also, we should provide the initial
+implementation that will be just redirecting all the requests to etcd, similarly
+to what we are doing today. This simple implementation should be used to enable
+easy transition to the new mechanism later (api unification).
+
+3. The target implementation internally will have a single watch to etcd open,
+responsible for watching all the changes of objects of a given resource type.
+However, this watch can potentially expire at any time and reconnecting can
+return "too old resource version". In that case relisting is necessary.
+To avoid LIST requests coming from all the watcher at that time, we will
+introduce an additional etcd event type:
+[EtcdResync](../../pkg/tools/etcd_helper_watch.go#L36)
```

Comment:
I'd really strongly suggest that you do this in a second step, not at the same time. And IMO this needs to be gated by API version. We can't just start sending this, it will break all v1 clients.

For the first step, just disconnect everyone.

This case should be super rare, as it should be easy for apiserver to keep up, so I'm not worried about it.